### ISSUE TITLE
BLD/IO: Elusive Travis ResourceWarning & closing files by default 

### ISSUE THREAD

**Issue by mroeschke** [pandas-dev/pandas#22675]
Every once in a while Travis will raise a `ResourceWarning`

```
____________________ TestPythonParser.test_no_header_prefix ____________________
[gw0] darwin -- Python 3.5.6 /Users/travis/miniconda3/envs/pandas/bin/python
self = <pandas.tests.io.parser.test_parsers.TestPythonParser object at 0x11d030da0>
    def test_no_header_prefix(self):
        data = """1,2,3,4,5
    6,7,8,9,10
    11,12,13,14,15
    """
        df_pref = self.read_table(StringIO(data), sep=',', prefix='Field'
>                                 header=None)
pandas/tests/io/parser/header.py:48:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas/tests/io/parser/test_parsers.py:111: in read_table
    df = read_table(*args, **kwds)
../../../miniconda3/envs/pandas/lib/python3.5/contextlib.py:66: in __exit__
    next(self.gen)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
E           AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning("unclosed file <_io.BufferedReader name='/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/tmpp8reunuatest_file.zip'>",), '/Users/travis/miniconda3/envs/pandas/lib/python3.5/site-packages/py/_vendored_packages/apipkg.py', 146)].
```

And currently our file closing mechanism for parser functions is like so:
https://github.com/pandas-dev/pandas/blob/73dd6ece3b66008eb8077098c101d0e300b1e714/pandas/io/parsers.py#L459-L463

It seems pointless having a generic `try`/`except` within an `if` statement (why have the if statement at all then?). More broadly, should we attempt to close all file-like objects anyways even if they are not necessarily needed (like StringIO in this Travis case)? Might help alleviate this Travis error.

Some examples:

- https://travis-ci.org/pandas-dev/pandas/jobs/427395352#L2258
- https://travis-ci.org/pandas-dev/pandas/jobs/427392469#L2259

**Comment by h-vetinari** [comment]
Kept having lots of those as well, and commented about it in the closest-seeming issue I found (https://github.com/pandas-dev/pandas/issues/13962#issuecomment-405025631), including some links to failed builds (July 2018).

**Comment by jbrockmendel** [comment]
> So an unrelated test is seemingly leaking a warning

Is it the warning that's coming from the unrelated test or just the unclosed file?

**Comment by TomAugspurger** [comment]
I was able to reproduce locally with `pytest -v -n 2 -m "not single" -r xX pandas/tests/io --count=5 `. So that's something....

> Is it the warning that's coming from the unrelated test or just the unclosed file?

It seems like the unclosed file warning has to be from a separate test right? In e.g. https://travis-ci.org/pandas-dev/pandas/jobs/427395352#L2258 we shouldn't be opening / closing a zip file.

**Comment by jbrockmendel** [comment]
> It seems like the unclosed file warning has to be from a separate test right? 

Yah, I'm just trying to rule out the extra-weird case where the ResourceWarning is being issued during the compression test but only showing up in the parser test, as opposed to the less-weird case where the compression test leaves a file open but that file only leads to a ResourceWarning later on.

Could check `psutil.open_files()` at the end of the compression test?

**Comment by TomAugspurger** [comment]
Gotcha, I suspect (hope) that the less-weird case is what's happening.

**Comment by TomAugspurger** [comment]
Hmm, this looks fishy: https://github.com/pandas-dev/pandas/blob/master/pandas/tests/io/test_pickle.py#L345-L348 We're creating a zipfile and not closing that. Let's give that a shot.

**Comment by TomAugspurger** [comment]
This is seemingly not fixed by #22679.

https://circleci.com/gh/pandas-dev/pandas/20081?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

```
__________________ TestPythonParser.test_dtype_with_converter __________________

self = <pandas.tests.io.parser.test_parsers.TestPythonParser object at 0x7f85c29ee2e8>

    def test_dtype_with_converter(self):
        data = """a,b
    1.1,2.2
    1.2,2.3"""
        # dtype spec ignored if converted specified
        with tm.assert_produces_warning(ParserWarning):
            result = self.read_csv(StringIO(data), dtype={'a': 'i8'},
>                                  converters={'a': lambda x: str(x)})

pandas/tests/io/parser/dtypes.py:340: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/conda/envs/pandas/lib/python3.5/contextlib.py:66: in __exit__
    next(self.gen)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected_warning = <class 'pandas.errors.ParserWarning'>
filter_level = 'always', clear = None, check_stacklevel = True

    @contextmanager
    def assert_produces_warning(expected_warning=Warning, filter_level="always",
                                clear=None, check_stacklevel=True):
        """
        Context manager for running code expected to either raise a specific
        warning, or not raise any warnings. Verifies that the code raises the
        expected warning, and that it does not raise any other unexpected
        warnings. It is basically a wrapper around ``warnings.catch_warnings``.
    
        Parameters
        ----------
        expected_warning : {Warning, False, None}, default Warning
            The type of Exception raised. ``exception.Warning`` is the base
            class for all warnings. To check that no warning is returned,
            specify ``False`` or ``None``.
        filter_level : str, default "always"
            Specifies whether warnings are ignored, displayed, or turned
            into errors.
            Valid values are:
    
            * "error" - turns matching warnings into exceptions
            * "ignore" - discard the warning
            * "always" - always emit a warning
            * "default" - print the warning the first time it is generated
              from each location
            * "module" - print the warning the first time it is generated
              from each module
            * "once" - print the warning the first time it is generated
    
        clear : str, default None
            If not ``None`` then remove any previously raised warnings from
            the ``__warningsregistry__`` to ensure that no warning messages are
            suppressed by this context manager. If ``None`` is specified,
            the ``__warningsregistry__`` keeps track of which warnings have been
            shown, and does not show them again.
        check_stacklevel : bool, default True
            If True, displays the line that called the function containing
            the warning to show were the function is called. Otherwise, the
            line that implements the function is displayed.
    
        Examples
        --------
        >>> import warnings
        >>> with assert_produces_warning():
        ...     warnings.warn(UserWarning())
        ...
        >>> with assert_produces_warning(False):
        ...     warnings.warn(RuntimeWarning())
        ...
        Traceback (most recent call last):
            ...
        AssertionError: Caused unexpected warning(s): ['RuntimeWarning'].
        >>> with assert_produces_warning(UserWarning):
        ...     warnings.warn(RuntimeWarning())
        Traceback (most recent call last):
            ...
        AssertionError: Did not see expected warning of class 'UserWarning'.
    
        ..warn:: This is *not* thread-safe.
        """
        with warnings.catch_warnings(record=True) as w:
    
            if clear is not None:
                # make sure that we are clearing these warnings
                # if they have happened before
                # to guarantee that we will catch them
                if not is_list_like(clear):
                    clear = [clear]
                for m in clear:
                    try:
                        m.__warningregistry__.clear()
                    except Exception:
                        pass
    
            saw_warning = False
            warnings.simplefilter(filter_level)
            yield w
            extra_warnings = []
    
            for actual_warning in w:
                if (expected_warning and issubclass(actual_warning.category,
                                                    expected_warning)):
                    saw_warning = True
    
                    if check_stacklevel and issubclass(actual_warning.category,
                                                       (FutureWarning,
                                                        DeprecationWarning)):
                        from inspect import getframeinfo, stack
                        caller = getframeinfo(stack()[2][0])
                        msg = ("Warning not set with correct stacklevel. "
                               "File where warning is raised: {actual} != "
                               "{caller}. Warning message: {message}"
                               ).format(actual=actual_warning.filename,
                                        caller=caller.filename,
                                        message=actual_warning.message)
                        assert actual_warning.filename == caller.filename, msg
                else:
                    extra_warnings.append((actual_warning.category.__name__,
                                           actual_warning.message,
                                           actual_warning.filename,
                                           actual_warning.lineno))
            if expected_warning:
                msg = "Did not see expected warning of class {name!r}.".format(
                    name=expected_warning.__name__)
                assert saw_warning, msg
            assert not extra_warnings, ("Caused unexpected warning(s): {extra!r}."
>                                       ).format(extra=extra_warnings)
E           AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning("unclosed file <_io.BufferedReader name='/tmp/tmp4hjpmwyntest_file.zip'>",), '/opt/conda/envs/pandas/lib/python3.5/importlib/_bootstrap.py', 192)].
```

**Comment by mroeschke** [comment]
Interesting, this error appears on a different test.

Does `zip_file` here need to be closed as well?
https://github.com/pandas-dev/pandas/blob/71852da03994c7c79a4ba3a0f91c6d723be6a299/pandas/util/testing.py#L195-L209

**Comment by TomAugspurger** [comment]
Hmm, that may be it. Let me play with that.

**Comment by TomAugspurger** [comment]
No luck on that I think. Closing the archive doesn't seem to have any effect if the file is closed. 

```
$ python -W "error::ResourceWarning"
Python 3.6.5 (default, Mar 30 2018, 06:41:53)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import zipfile
>>> zf = zipfile.ZipFile("test.zip")
>>> f = zf.open(zf.namelist()[0])
>>>
Exception ignored in: <_io.FileIO name='test.zip' mode='rb' closefd=True>
ResourceWarning: unclosed file <_io.BufferedReader name='test.zip'>
```



```
$ python -W "error::ResourceWarning"
Python 3.6.5 (default, Mar 30 2018, 06:41:53)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import zipfile
>>> zf = zipfile.ZipFile("test.zip")
>>> f = zf.open(zf.namelist()[0])
>>> f.close()
>>>
```

**Comment by h-vetinari** [comment]
Logging another case of this:

https://travis-ci.org/pandas-dev/pandas/jobs/428987214

```
=================================== FAILURES ===================================
_______________ TestPythonParser.test_categorical_dtype_encoding _______________
[gw0] linux -- Python 3.7.0 /home/travis/miniconda3/envs/pandas/bin/python
self = <pandas.tests.io.parser.test_parsers.TestPythonParser object at 0x7fe1e0d04b00>
datapath = <function datapath.<locals>.deco at 0x7fe1d33a4488>
    def test_categorical_dtype_encoding(self, datapath):
        # GH 10153
        pth = datapath('io', 'parser', 'data', 'unicode_series.csv')
        encoding = 'latin-1'
        expected = self.read_csv(pth, header=None, encoding=encoding)
        expected[1] = Categorical(expected[1])
        actual = self.read_csv(pth, header=None, encoding=encoding
                               dtype={1: 'category'})
        tm.assert_frame_equal(actual, expected)
        pth = datapath('io', 'parser', 'data', 'utf16_ex.txt')
        encoding = 'utf-16'
        expected = self.read_table(pth, encoding=encoding)
        expected = expected.apply(Categorical)
>       actual = self.read_table(pth, encoding=encoding, dtype='category')
[...]
E           AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning("unclosed file <_io.BufferedReader name='/tmp/tmpevb83mjxtest_file.zip'>"), '/home/travis/miniconda3/envs/pandas/lib/python3.7/site-packages/numpy/core/numeric.py', 501)].
```

**Comment by mroeschke** [comment]
Closed by #22699 I believe.
